src/nvim/ops.c

```diff
@@ -1504,20 +1504,23 @@ int op_delete(oparg_T *oap)
     yankreg_T *reg = NULL;
     int did_yank = false;
     if (oap->regname != 0) {
-      // yank without message
-      did_yank = op_yank(oap, false, true);
-      if (!did_yank) {
-        // op_yank failed, don't do anything
+      // check for read-only register
+      if (!valid_yank_reg(oap->regname, true)) {
+        beep_flush();
         return OK;
       }
+      reg = get_yank_register(oap->regname, YREG_YANK);
+      // yank without message
+      op_yank_reg(oap, false, reg, is_append_register(oap->regname));
+      did_yank = true;
```

Comment:
I cannot keep the `op_yank()` call because after this patch, `reg` needs to be assigned here as it isn't always assigned a value below.